### PR TITLE
[api-minor] JS -- Add listener for sandbox events only if there are some actions

### PR DESCRIPTION
**Pull request by calixteman**
* When no actions then set it to null instead of empty object
* Even if a field has no actions, it needs to listen to events from the sandbox in order to be updated if an action changes something in it.

### CODE REVIEW

File: web/base_viewer.js [hunk 1]
```diff
@@ -497,8 +500,8 @@ class BaseViewer {
 
     // Fetch a single page so we can get a viewport that will be the default
     // viewport for all pages
-    firstPagePromise
-      .then(firstPdfPage => {
+    Promise.all([firstPagePromise, hasJSActionsPromise])
```
Comment (brendandahl) [inline web/base_viewer.js:503]: This section is pretty important to display things as fast as possible. Could we pass in the hasJSActionsPromise into the page view instead so it's resolved a a later time and we don't have to wait for it to starting showing pages?

**Review by brendandahl — CHANGES_REQUESTED** [review]
Looks good overall, but I'd like to hear if it's possible to address the comment.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Can't we actually delay this *completely* until Annotation parsing actually occurs, since in most documents there's either no annotations present or at least no JavaScript actions present!?

That'd probably reduce the overall size of this, and also simplify the implementation (especially on the viewer side). One quick idea that I had, was to simply include this property in the annotation data *itself* such that you'll only need to pass around the `enableScripting`-value to the various viewer-components.

To illustrate my idea, perhaps something along these lines (purposely incomplete, but only to illustrate my idea):
```diff
diff --git a/src/core/annotation.js b/src/core/annotation.js
index 252e1529b..f28806726 100644
--- a/src/core/annotation.js
+++ b/src/core/annotation.js
@@ -62,13 +62,17 @@ class AnnotationFactory {
    *   instance.
    */
   static create(xref, ref, pdfManager, idFactory) {
-    return pdfManager.ensureCatalog("acroForm").then(acroForm => {
+    return Promise.all([
+      pdfManager.ensureCatalog("acroForm"),
+      pdfManager.ensureDoc("hasJSActions"),
+    ]).then(([acroForm, hasJSActions]) => {
       return pdfManager.ensure(this, "_create", [
         xref,
         ref,
         pdfManager,
         idFactory,
         acroForm,
+        hasJSActions,
       ]);
     });
   }
@@ -76,7 +80,14 @@ class AnnotationFactory {
   /**
    * @private
    */
-  static _create(xref, ref, pdfManager, idFactory, acroForm) {
+  static _create(
+    xref,
+    ref,
+    pdfManager,
+    idFactory,
+    acroForm,
+    hasJSActions = false
+  ) {
     const dict = xref.fetchIfRef(ref);
     if (!isDict(dict)) {
       return undefined;
@@ -97,6 +108,7 @@ class AnnotationFactory {
       id,
       pdfManager,
       acroForm: acroForm instanceof Dict ? acroForm : Dict.empty,
+      hasJSActions,
     };
 
     switch (subtype) {
@@ -281,6 +293,7 @@ class Annotation {
       modificationDate: this.modificationDate,
       rect: this.rectangle,
       subtype: params.subtype,
+      documentHasJSActions: params.hasJSActions,
     };
 
     this._fallbackFontDict = null;
diff --git a/src/display/annotation_layer.js b/src/display/annotation_layer.js
index effa122b5..4e07676fe 100644
--- a/src/display/annotation_layer.js
+++ b/src/display/annotation_layer.js
@@ -478,7 +478,8 @@ class TextWidgetAnnotationElement extends WidgetAnnotationElement {
         event.target.setSelectionRange(0, 0);
       });
 
-      if (this.data.actions) {
+      // Assuming that the `enableScripting`-value is passed in:
+      if (this.enableScripting && this.data.documentHasJSActions) {
         element.addEventListener("updateFromSandbox", function (event) {
           const data = event.detail;
           if ("value" in data) {
```

**Comment by calixteman** [comment]
I think this is a very good idea: I like it, thank you.

**Review by brendandahl — APPROVED** [review]

File: src/core/annotation.js [hunk 2]
```diff
@@ -59,25 +59,39 @@ class AnnotationFactory {
    * @param {Object} ref
    * @param {PDFManager} pdfManager
    * @param {Object} idFactory
+   * @param {boolean} checkJSActions
    * @returns {Promise} A promise that is resolved with an {Annotation}
    *   instance.
    */
-  static create(xref, ref, pdfManager, idFactory) {
-    return pdfManager.ensureCatalog("acroForm").then(acroForm => {
+  static create(xref, ref, pdfManager, idFactory, checkJSActions = true) {
```
Comment (Snuffleupagus) [inline src/core/annotation.js:66]: Can't we make `false` the default value instead, and explicitly enable it where necessary?

File: src/core/annotation.js [hunk 3]
```diff
@@ -59,25 +59,39 @@ class AnnotationFactory {
...
+    return Promise.all([
+      pdfManager.ensureCatalog("acroForm"),
+      checkJSActions
+        ? pdfManager.ensureDoc("hasJSActions")
+        : Promise.resolve(false),
```
Comment (Snuffleupagus) [inline src/core/annotation.js:71]: I believe that the following ought to suffice, since non-Promise values should be automatically wrapped in a Promise:
```suggestion
      checkJSActions && pdfManager.ensureDoc("hasJSActions"),
```

File: src/core/document.js [hunk 4]
```diff
@@ -962,12 +962,14 @@ class PDFDocument {
     if (!(name in promises)) {
       promises.set(name, []);
     }
+
```
Comment (Snuffleupagus) [inline src/core/document.js:965]: Nit: Looks like an unrelated change!?

File: src/display/api.js [hunk 5]
```diff
@@ -903,6 +903,14 @@ class PDFDocumentProxy {
...
   }
 
+  /**
+   * @returns {Promise<Boolean>} A promise that is resolved with true if
+   * there are some actions in JavaScript.
+   */
+  hasJSActions() {
+    return this._transport.hasJSActions();
+  }
+
```
Comment (Snuffleupagus) [inline src/display/api.js:913]: It *looks* like this method is now completely unused/untested in the latest version of the patch.

Do we actually need to add this to the API, or could we instead wait until we see if there's an actual use for this? Generally speaking, it feels best to avoid adding unused things to the official API since we then have to maintain something that's unused/untested.

Hence I'd suggest reverting the changes in both the `src/display/api.js` and `src/core/worker.js` files for now.

File: web/base_viewer.js [hunk 6]
```diff
@@ -77,6 +77,7 @@ const DEFAULT_CACHE_SIZE = 10;
  *   total pixels, i.e. width * height. Use -1 for no limit. The default value
  *   is 4096 * 4096 (16 mega-pixels).
  * @property {IL10n} l10n - Localization service.
+ * @property {boolean} [enableScripting] - Enable JS evaluation.
```
Comment (Snuffleupagus) [inline web/base_viewer.js:80]: Please mention the default value here, and make sure to use the *same* description in both `BaseViewer` and `PDFPageView`, i.e.
```suggestion
 * @property {boolean} [enableScripting] - Enable embedded script execution.
 *   The default value is `false`.
```

File: web/pdf_page_view.js [hunk 7]
```diff
@@ -63,6 +63,7 @@ import { viewerCompatibilityParams } from "./viewer_compatibility.js";
  *   total pixels, i.e. width * height. Use -1 for no limit. The default value
  *   is 4096 * 4096 (16 mega-pixels).
  * @property {IL10n} l10n - Localization service.
+ * @property {boolean} enableScripting - Enable embedded script execution.
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:66]: Please mention the default value here as well:
```javascript
 *   The default value is `false`.
```
`*   The default value is `false`.`

File: src/display/annotation_layer.js [hunk 8]
```diff
@@ -1565,6 +1568,7 @@ class AnnotationLayer {
         svgFactory: new DOMSVGFactory(),
         annotationStorage:
           parameters.annotationStorage || new AnnotationStorage(),
+        enableScripting: parameters.enableScripting,
```
Comment (Snuffleupagus) [inline src/display/annotation_layer.js:1571]: You also need to update the `AnnotationElementParameters`-typedef at the top of this file, hence please add the following there:
```javascript
 * @property {boolean} [enableScripting] - Enable embedded script execution.
 *   The default value is `false`.
```

**Review by Snuffleupagus — COMMENTED** [review]
Looks much better this way; I've added a few comments, after which this should be good to go (assuming all tests pass of course :-)

File: src/core/document.js [hunk 9]
```diff
@@ -1020,6 +1021,26 @@ class PDFDocument {
...
+  get hasJSActions() {
+    return shadow(
+      this,
+      "hasJSActions",
+      this.fieldObjects.then(objects => {
+        if (objects === null) {
+          return false;
+        }
+        for (const object of Object.values(objects)) {
+          for (const obj of object) {
```
Comment (timvandermeij) [inline src/core/document.js:1033]: It might be more clear to rename `objects` -> `fieldObjects`, `object` -> `fieldObject` and `obj` -> `object`.

**Review by timvandermeij — COMMENTED** [review]
Looks good but I'd really like to see the property comment addressed since it doesn't feel like the right place to put it.

File: src/core/annotation.js [hunk 10]
```diff
@@ -282,6 +295,7 @@ class Annotation {
       modificationDate: this.modificationDate,
       rect: this.rectangle,
       subtype: params.subtype,
+      documentHasJSActions: params.hasJSActions,
```
[the remark this answers is not in the document]
Comment (Snuffleupagus) [reply src/core/annotation.js]: > I don't really see why we need to and would want to store a document-wide property on every individual annotation, especially since this is then also an API change. 

Sorry, but how is adding a *new* boolean to the data returned by `getAnnotations` a problem? :-)
The data in question is also cached, on the worker-side, so that shouldn't be much of an issue.

The reason why I suggested the previous approach, is that doing it the way you suggest (and what the patch originally did), has *really* undesirable consequences for the viewer since that first of all gives a number of the viewer-components (e.g. `PDFPageView`) *very* inconsistent API-surfaces!
Note also the point below, about unnecessary things API-calls running too early.

> It would be much better to use the same logic as enableScripting, which is also a document-wide property, where we calculate the property once and pass it into the AnnotationLayer class.

Sorry, but I don't agree with this at all, since it ends up forcing an unnecessary API-call early on, *before* rendering has started. That's why I suggested the previous approach, since it delays these things until the Annotations are actually fetched.
You absolutely want to avoid *all* strictly unnecessary API-calls *before* rendering has started!

---

Given that we apparently have very different positions/perspectives here, please give me some time to suggest a compromise here!

File: web/pdf_page_view.js [hunk 11]
```diff
@@ -541,26 +547,30 @@ class PDFPageView {
...
+      div.setAttribute("data-loaded", true);
 
-    this.eventBus.dispatch("pagerender", {
-      source: this,
-      pageNumber: this.id,
+      this.eventBus.dispatch("pagerender", {
+        source: this,
+        pageNumber: this.id,
+      });
+      return resultPromise;
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:572]: You *absolutely* can't delay this event, since that will affect things like clean-up (and thus resource usage) in the viewer negatively.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: src/core/annotation.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Snuffleupagus) [reply src/core/annotation.js]: I'd suggest the following approach, to ensure that `hasJSActions` isn't called *before* rendering has started and avoid `BaseViewer`/`PDFPageView` diverging unnecessarily:
```diff
diff --git a/src/display/annotation_layer.js b/src/display/annotation_layer.js
index 705568575..b8e1c90fe 100644
--- a/src/display/annotation_layer.js
+++ b/src/display/annotation_layer.js
@@ -43,6 +43,8 @@ import { AnnotationStorage } from "./annotation_storage.js";
  *   for annotation icons. Include trailing slash.
  * @property {boolean} renderInteractiveForms
  * @property {Object} svgFactory
+ * @property {boolean} [enableScripting]
+ * @property {boolean} [hasJSActions]
  */
 
 class AnnotationElementFactory {
diff --git a/src/display/api.js b/src/display/api.js
index 8ffcc1495..b58830463 100644
--- a/src/display/api.js
+++ b/src/display/api.js
@@ -904,8 +904,8 @@ class PDFDocumentProxy {
   }
 
   /**
-   * @returns {Promise<boolean>} A promise that is resolved with true
-   * if some fields have some JS actions.
+   * @returns {Promise<boolean>} A promise that is resolved with `true` if some
+   *   /AcroForm fields has JavaScript actions.
    */
   hasJSActions() {
     return this._transport.hasJSActions();
diff --git a/web/annotation_layer_builder.js b/web/annotation_layer_builder.js
index e37ee831f..c4d47112a 100644
--- a/web/annotation_layer_builder.js
+++ b/web/annotation_layer_builder.js
@@ -28,7 +28,8 @@ import { SimpleLinkService } from "./pdf_link_service.js";
  * @property {IPDFLinkService} linkService
  * @property {DownloadManager} downloadManager
  * @property {IL10n} l10n - Localization service.
- * @property {boolean} enableScripting
+ * @property {boolean} [enableScripting]
+ * @property {boolean} [hasJSActionsPromise]
  */
 
 class AnnotationLayerBuilder {
@@ -45,7 +46,7 @@ class AnnotationLayerBuilder {
     renderInteractiveForms = true,
     l10n = NullL10n,
     enableScripting = false,
-    hasJSActions = false,
+    hasJSActionsPromise = null,
   }) {
     this.pageDiv = pageDiv;
     this.pdfPage = pdfPage;
@@ -56,7 +57,7 @@ class AnnotationLayerBuilder {
     this.l10n = l10n;
     this.annotationStorage = annotationStorage;
     this.enableScripting = enableScripting;
-    this.hasJSActions = hasJSActions;
+    this._hasJSActionsPromise = hasJSActionsPromise;
 
     this.div = null;
     this._cancelled = false;
@@ -69,7 +70,10 @@ class AnnotationLayerBuilder {
    *   annotations is complete.
    */
   render(viewport, intent = "display") {
-    return this.pdfPage.getAnnotations({ intent }).then(annotations => {
+    return Promise.all([
+      this.pdfPage.getAnnotations({ intent }),
+      this._hasJSActionsPromise,
+    ]).then(([annotations, hasJSActions = false]) => {
       if (this._cancelled) {
         return;
       }
@@ -88,7 +92,7 @@ class AnnotationLayerBuilder {
         downloadManager: this.downloadManager,
         annotationStorage: this.annotationStorage,
         enableScripting: this.enableScripting,
-        hasJSActions: this.hasJSActions,
+        hasJSActions,
       };
 
       if (this.div) {
@@ -133,8 +137,8 @@ class DefaultAnnotationLayerFactory {
    *   for annotation icons. Include trailing slash.
    * @param {boolean} renderInteractiveForms
    * @param {IL10n} l10n
-   * @param {boolean} enableScripting
-   * @param {boolean} hasJSActions
+   * @param {boolean} [enableScripting]
+   * @param {boolean} [hasJSActionsPromise]
    * @returns {AnnotationLayerBuilder}
    */
   createAnnotationLayerBuilder(
@@ -145,7 +149,7 @@ class DefaultAnnotationLayerFactory {
     renderInteractiveForms = true,
     l10n = NullL10n,
     enableScripting = false,
-    hasJSActions = false
+    hasJSActionsPromise = null
   ) {
     return new AnnotationLayerBuilder({
       pageDiv,
@@ -156,7 +160,7 @@ class DefaultAnnotationLayerFactory {
       l10n,
       annotationStorage,
       enableScripting,
-      hasJSActions,
+      hasJSActionsPromise,
     });
   }
 }
diff --git a/web/base_viewer.js b/web/base_viewer.js
index cf2c8552a..966eaf342 100644
--- a/web/base_viewer.js
+++ b/web/base_viewer.js
@@ -468,7 +468,6 @@ class BaseViewer {
 
     const annotationStorage = pdfDocument.annotationStorage;
     const optionalContentConfigPromise = pdfDocument.getOptionalContentConfig();
-    const hasJSActionsPromise = pdfDocument.hasJSActions();
 
     this._pagesCapability.promise.then(() => {
       this.eventBus.dispatch("pagesloaded", {
@@ -532,7 +531,6 @@ class BaseViewer {
             maxCanvasPixels: this.maxCanvasPixels,
             l10n: this.l10n,
             enableScripting: this.enableScripting,
-            hasJSActionsPromise,
           });
           this._pages.push(pageView);
         }
@@ -1214,8 +1212,7 @@ class BaseViewer {
    *   for annotation icons. Include trailing slash.
    * @param {boolean} renderInteractiveForms
    * @param {IL10n} l10n
-   * @param {boolean} enableScripting
-   * @param {boolean} hasJSActions
+   * @param {boolean} [enableScripting]
    * @returns {AnnotationLayerBuilder}
    */
   createAnnotationLayerBuilder(
@@ -1225,8 +1222,7 @@ class BaseViewer {
     imageResourcesPath = "",
     renderInteractiveForms = false,
     l10n = NullL10n,
-    enableScripting = false,
-    hasJSActions = false
+    enableScripting = false
   ) {
     return new AnnotationLayerBuilder({
       pageDiv,
@@ -1238,7 +1234,7 @@ class BaseViewer {
       downloadManager: this.downloadManager,
       l10n,
       enableScripting,
-      hasJSActions,
+      hasJSActionsPromise: enableScripting && this.pdfDocument.hasJSActions(),
     });
   }
 
diff --git a/web/interfaces.js b/web/interfaces.js
index 062369362..41a4fe571 100644
--- a/web/interfaces.js
+++ b/web/interfaces.js
@@ -181,8 +181,8 @@ class IPDFAnnotationLayerFactory {
    *   for annotation icons. Include trailing slash.
    * @param {boolean} renderInteractiveForms
    * @param {IL10n} l10n
-   * @param {boolean} enableScripting
-   * @param {boolean} hasJSActions
+   * @param {boolean} [enableScripting]
+   * @param {boolean} [hasJSActionsPromise]
    * @returns {AnnotationLayerBuilder}
    */
   createAnnotationLayerBuilder(
@@ -193,7 +193,7 @@ class IPDFAnnotationLayerFactory {
     renderInteractiveForms = true,
     l10n = undefined,
     enableScripting = false,
-    hasJSActions = false
+    hasJSActionsPromise = null
   ) {}
 }
 
diff --git a/web/pdf_page_view.js b/web/pdf_page_view.js
index 2d28d9d8c..72b338764 100644
--- a/web/pdf_page_view.js
+++ b/web/pdf_page_view.js
@@ -63,10 +63,8 @@ import { viewerCompatibilityParams } from "./viewer_compatibility.js";
  *   total pixels, i.e. width * height. Use -1 for no limit. The default value
  *   is 4096 * 4096 (16 mega-pixels).
  * @property {IL10n} l10n - Localization service.
- * @property {boolean} enableScripting - Enable embedded script execution.
+ * @property {boolean} [enableScripting] - Enable embedded script execution.
  *   The default value is `false`.
- * @property {Promise<boolean>} hasJSActionsPromise - A promise that is resolved
- * with true if some fields have some JS actions.
  */
 
 const MAX_CANVAS_PIXELS = viewerCompatibilityParams.maxCanvasPixels || 16777216;
@@ -114,7 +112,6 @@ class PDFPageView {
     this.enableWebGL = options.enableWebGL || false;
     this.l10n = options.l10n || NullL10n;
     this.enableScripting = options.enableScripting || false;
-    this.hasJSActionsPromise = options.hasJSActionsPromise;
 
     this.paintTask = null;
     this.paintedViewportMap = new WeakMap();
@@ -547,30 +544,27 @@ class PDFPageView {
       }
     );
 
-    return this.hasJSActionsPromise.then(hasJSActions => {
-      if (this.annotationLayerFactory) {
-        if (!this.annotationLayer) {
-          this.annotationLayer = this.annotationLayerFactory.createAnnotationLayerBuilder(
-            div,
-            pdfPage,
-            this._annotationStorage,
-            this.imageResourcesPath,
-            this.renderInteractiveForms,
-            this.l10n,
-            this.enableScripting,
-            hasJSActions
-          );
-        }
-        this._renderAnnotationLayer();
+    if (this.annotationLayerFactory) {
+      if (!this.annotationLayer) {
+        this.annotationLayer = this.annotationLayerFactory.createAnnotationLayerBuilder(
+          div,
+          pdfPage,
+          this._annotationStorage,
+          this.imageResourcesPath,
+          this.renderInteractiveForms,
+          this.l10n,
+          this.enableScripting
+        );
       }
-      div.setAttribute("data-loaded", true);
+      this._renderAnnotationLayer();
+    }
+    div.setAttribute("data-loaded", true);
 
-      this.eventBus.dispatch("pagerender", {
-        source: this,
-        pageNumber: this.id,
-      });
-      return resultPromise;
+    this.eventBus.dispatch("pagerender", {
+      source: this,
+      pageNumber: this.id,
     });
+    return resultPromise;
   }
 
   paintOnCanvas(canvasWrapper) {
```

[the remark this answers is not in the document]
Comment (timvandermeij) [reply src/core/annotation.js]: Oh, I'm perfectly fine with this approach of delaying `hasJSActions`. My concerns here were only with the changes in `src/core/annotation.js` since it doesn't feel right to add a document-wide property on annotation instances (not the boolean itself being the problem, but merely the location of the data being out of place). Annotation instances should only contain information about the annotations themselves, and `hasJSActions` is not a property of the annotation but of the document. A solution along the lines of `enableScripting`, which also isn't included in every annotation instance, was merely a direction that could be thought about to move that data out of the annotation instances, but that wasn't really clear from my side; sorry for that.

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Looks good to me too now. Thank you!
